src/lib/logical_query_plan/join_node.cpp

```diff
@@ -228,19 +230,72 @@ bool JoinNode::is_column_nullable(const ColumnID column_id) const {
 
 const std::vector<std::shared_ptr<AbstractExpression>>& JoinNode::join_predicates() const { return node_expressions; }
 
+void JoinNode::mark_as_semi_reduction(const std::shared_ptr<JoinNode>& reduced_join_node) {
+  Assert(!_is_semi_reduction, "The semi reducer status should be set once only.");
+  Assert(reduced_join_node, "Reduced JoinNode must be provided.");
+  Assert(join_mode == JoinMode::Semi, "Semi join reductions require JoinMode::Semi.");
+  DebugAssert(join_predicates().size() == 1,
+              "Currently, semi join reductions are expected to have a single join predicate.");
+  DebugAssert(std::any_of(reduced_join_node->join_predicates().begin(), reduced_join_node->join_predicates().end(),
+                          [&](const auto predicate) { return *predicate == *join_predicates()[0]; }),
+              "Both semi join reduction node and the reduced join should have a common join predicate.");
+  _is_semi_reduction = true;
+  _reduced_join_node = std::weak_ptr<JoinNode>(reduced_join_node);
+}
+
+bool JoinNode::is_semi_reduction() const {
+  DebugAssert(!_is_semi_reduction || join_mode == JoinMode::Semi, "Non-semi join is marked as a semi reduction.");
+  return _is_semi_reduction;
+}
+
+std::shared_ptr<JoinNode> JoinNode::get_or_find_reduced_join_node() const {
+  if (!_is_semi_reduction) {
+    return nullptr;
+  }
+
+  if (_reduced_join_node.expired()) {
+    // In deep copies of the LQP, the weak pointer to the reduced join is expired (lazy discovery). In such cases,
+    // find the reduced join by traversing the LQP upwards.
+    visit_lqp_upwards(std::const_pointer_cast<AbstractLQPNode>(shared_from_this()), [&](const auto& current_node) {
+      if (current_node->type != LQPNodeType::Join || current_node.get() == this) {
+        return LQPUpwardVisitation::VisitOutputs;
+      }
+      const auto join_node = std::static_pointer_cast<JoinNode>(current_node);
+      if (std::none_of(join_node->join_predicates().begin(), join_node->join_predicates().end(),
+                       [&](const auto& predicate) { return *predicate == *join_predicates()[0]; })) {
+        return LQPUpwardVisitation::VisitOutputs;
+      }
+
+      _reduced_join_node = std::weak_ptr<JoinNode>(join_node);
+      return LQPUpwardVisitation::DoNotVisitOutputs;
+    });
+
+    Assert(!_reduced_join_node.expired(), "Could not find JoinNode that gets reduced by this semi join reduction.");
+  }
+
+  return _reduced_join_node.lock();
+}
+
 size_t JoinNode::_on_shallow_hash() const { return boost::hash_value(join_mode); }
```

Comment:
What implicatons arise from ignoring the semi join flag in the hash and equality check for hash-based containers? Are we sure that there are not false matches?

Reply:
Hmm. As a consequence, we will not be able to add two different semi join node instances to a hash-based container, if their only difference is the semi reduction flag property.

Reply:
Having the same semi-join (intendedly) twice in the LQP with and without reduction flag should be pretty rare, too. So I'd also accept that they replace each other in a hash container if we document it somewhere. Maybe some different opinions from @Bouncner or @mweisgut?

Reply:
But in this scenario, the visualization of the resulting physical query plan would show incorrect information: Since we translate in a bottom-up way, we would translate the reducer node first. When the reduced node is translated, we would find a match in `_operator_by_lqp_node` and use the stored join operator which points to the reducer JoinNode. If we call `description` for the visualization, we would evaluate the JoiNode as semi reducer and return `(Semi Reduction)` - still preferring `(Semi Reducer)`. Thus, both the reducer and reduced nodes would both show as reducer nodes, wouldn't they?

Reply:
The reduced JoinNode should have at least one different input node because it follows the semi reduction node.

Anyway, I add the attributes to shallow hash and equals.

Reply:
You mean sth like
```c++
const auto reducer_node = JoinNode::make(JoinMode::Semi, equals(_b_a, _a_a)),
  _node_a,
  _node_b));

const auto inner_join = JoinNode::make(JoinMode::Inner, redcuer_node, _node_b);
reducer_node->mark_as_semi_reduction(inner_join);
```

vs a dedicated constructor? How often do we have this problem in the tests? I wouldn't say we need a new join constructor for just a few test cases

Reply:
I also vote for not adding a new constructor for a few test cases.

What about sticking to the `expected_lqp` version, removing const, and calling `mark_as_semi_reduction(expected_lqp)` for the expected lqp's left input node?

Reply:
even shorter, I think that's fine

Reply:
> What about sticking to the expected_lqp version, removing const, and calling mark_as_semi_reduction(expected_lqp) for the expected lqp's left input node?

> then we need to cast expected_lqp->left_input() as join node to call the marker function, right?

Exactly. It works but adds some boilerplate code to the tests. In this PR, it only affects 4 tests. Some more will come with the SemiJoinReductionRemovalRule.